compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/CallAndReferenceGenerator.kt

```diff
@@ -361,14 +361,35 @@ class CallAndReferenceGenerator(
         }.applyCallArguments(annotationCall, annotationMode = true)
     }
 
-    fun convertToGetObject(qualifier: FirResolvedQualifier): IrExpression {
-        return convertToGetObject(qualifier, callableReferenceMode = false)!!
+    internal fun convertToGetObject(qualifier: FirResolvedQualifier): IrExpression {
+        return convertToGetObject(qualifier, null)!!
     }
 
-    internal fun convertToGetObject(qualifier: FirResolvedQualifier, callableReferenceMode: Boolean): IrExpression? {
+    internal fun convertToGetObject(
+        qualifier: FirResolvedQualifier,
+        callableReferenceAccess: FirCallableReferenceAccess? = null
+    ): IrExpression? {
         val classSymbol = (qualifier.typeRef.coneType as? ConeClassLikeType)?.lookupTag?.toSymbol(session)
-        if (callableReferenceMode && classSymbol is FirRegularClassSymbol) {
-            if (classSymbol.classId != qualifier.classId) {
+        if (callableReferenceAccess != null && classSymbol is FirRegularClassSymbol) {
+            val classIdMatched = classSymbol.classId == qualifier.classId
+            val companionClassIdMatched = classSymbol.fir.isCompanion && classSymbol.classId.outerClassId == qualifier.classId
+            val lastClassIfCompanion = conversionScope.lastClass()?.takeIf { it.isCompanion }
+            val callable =
+                ((callableReferenceAccess.calleeReference as FirResolvedNamedReference).resolvedSymbol as FirCallableSymbol<*>)
+            val companionMemberReferenceThatNeedsDispatchReceiver = companionClassIdMatched &&
+                    // TODO: this can't cover delegated members. See bb test kt38664.kt
+                    // Make sure the reference indeed refers to a member of that companion
+                    classSymbol.fir.declarations.contains(callable.fir) &&
+                    // Built-in extensions are stored in the companion, e.g., String.Companion.
+                    // Make sure this is not a reference to an extension function.
+                    !callable.isExtension &&
+                    // Dispatch receiver is not needed if we're converting a callable reference inside that companion, e.g.,
+                    // class Foo {
+                    //   companion object {
+                    //      val cp = Foo::member
+                    // } }
+                    lastClassIfCompanion?.classId != classSymbol.classId
```

Comment:
After some consideration I dropped this check. It looks very inconsistent, and does not help any BB / text test to pass.

Reply:
It helped prevent any regressions, but probably those regressions were due to missed parent check you mentioned below. Thanks.